dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/DefaultIconService.java

```diff
@@ -27,63 +27,226 @@
+    public Collection<Icon> getIcons( String[] keywords )
     {
-        return icons.values().stream()
-            .filter( icon -> Arrays.asList( icon.getKeywords() ).containsAll( keywords ) )
-            .collect( Collectors.toList() );
+        return Stream.concat( defaultIcons.values().stream()
+            .filter( icon -> Set.of( icon.getKeywords() ).containsAll( List.of( keywords ) ) ).toList().stream(),
+            customIconStore.getIconsByKeywords( keywords ).stream() )
+            .toList();
     }
 
     @Override
-    public Optional<IconData> getIcon( String key )
+    @Transactional( readOnly = true )
+    public Icon getIcon( String key )
+        throws NotFoundException
     {
-        return Optional.ofNullable( icons.get( key ) );
+        if ( defaultIcons.containsKey( key ) )
+        {
+            return defaultIcons.get( key );
+        }
+        else
+        {
+            CustomIcon customIcon = customIconStore.getIconByKey( key );
+            if ( customIcon == null )
+            {
+                throw new NotFoundException( String.format( "Icon not found: %s", key ) );
+            }
+
+            return customIcon;
+        }
+    }
+
+    @Override
+    @Transactional( readOnly = true )
+    public CustomIcon getCustomIcon( String key )
+        throws NotFoundException
+    {
+        CustomIcon customIcon = customIconStore.getIconByKey( key );
+        if ( customIcon == null )
+        {
+            throw new NotFoundException( String.format( "Custom icon not found: %s", key ) );
+        }
+
+        return customIcon;
     }
 
     @Override
     public Optional<Resource> getIconResource( String key )
+        throws NotFoundException
+    {
+        if ( defaultIcons.containsKey( key ) )
+        {
+            return Optional
+                .of( new ClassPathResource( String.format( "%s/%s.%s", ICON_PATH, key, DefaultIcon.Icons.SUFFIX ) ) );
+        }
+
+        throw new NotFoundException( String.format( "No default icon found with key %s.", key ) );
+    }
+
+    @Override
+    @Transactional( readOnly = true )
+    public Set<String> getKeywords()
+    {
+        return Stream.concat( defaultIcons.values().stream()
+            .map( Icon::getKeywords )
+            .flatMap( Arrays::stream ), customIconStore.getKeywords().stream() ).collect( Collectors.toSet() );
+    }
+
+    @Override
+    @Transactional( readOnly = true )
+    public boolean iconExists( String key )
+    {
+        return defaultIcons.get( key ) != null || customIconStore.getIconByKey( key ) != null;
+    }
+
+    @Override
+    @Transactional
+    public void addCustomIcon( CustomIcon customIcon )
+        throws BadRequestException,
+        NotFoundException
+    {
+        validateIconExists( customIcon.getKey() );
+        FileResource fileResource = getFileResource( customIcon.getFileResourceUid() );
+        customIconStore.save( customIcon, fileResource.getId(), currentUserService.getCurrentUser().getId() );
```

Comment:
In my opinion `id` is a repository concept and we shouldn't have to deal with it in the service.
I would change the signature of the `save` method to either accept `FileResource` and `User` object and extract the id itself or to just get the `customIcon` and through the `uid` join with the relevant tables and get the needed `id`s